src/components/inputs/number-input/number-input.spec.js

```diff
@@ -68,203 +64,96 @@ describe('NumberInput.hasFractionDigits', () => {
+    expect(container.querySelector('input')).toHaveClass('warning');
+  });
 
-        it('should have ARIA properties for the readonly state', () => {
-          expect(input).toHaveProp('aria-readonly', true);
-        });
-      });
-    });
+  it('should have error style when error is present', () => {
+    const { container } = render(<NumberInput {...baseProps} hasError />);
+    expect(container.querySelector('input')).toHaveClass('error');
   });
-});
 
-describe('callbacks', () => {
-  describe('when changing value', () => {
-    let wrapper;
-    let props;
-    let input;
-    beforeEach(() => {
-      props = createTestProps({
-        value: '1',
-        onChange: jest.fn(),
-      });
-      wrapper = shallow(<NumberInput {...props} />);
-      input = wrapper.find('input');
-      input.simulate('change', { target: { value: '2' } });
-    });
+  it('should have an HTML name', () => {
+    const { container } = render(<NumberInput {...baseProps} name="foo" />);
+    expect(container.querySelector('input')).toHaveAttribute('name', 'foo');
+  });
 
-    it('should call onChange', () => {
-      expect(props.onChange).toHaveBeenCalled();
-    });
+  it('should have ARIA role', () => {
+    const { container } = render(<NumberInput {...baseProps} />);
+    expect(container.querySelector('input')).toHaveAttribute('role', 'textbox');
+  });
 
-    it('should update with new value', () => {
-      expect(props.onChange).toHaveBeenCalledWith({
-        target: {
-          value: '2',
-        },
-      });
-    });
+  it('should forward the passed value', () => {
+    const { container } = render(<NumberInput {...baseProps} value="foo" />);
+    expect(container.querySelector('input')).toHaveAttribute('value', 'foo');
   });
-  describe('when input gains focus', () => {
-    let wrapper;
-    let props;
-    let input;
-    beforeEach(() => {
-      props = createTestProps({
-        value: '1',
-        onFocus: jest.fn(),
-      });
-      wrapper = shallow(<NumberInput {...props} />);
-      input = wrapper.find('input');
-      input.simulate('focus');
-    });
 
-    it('should call onFocus', () => {
-      expect(props.onFocus).toHaveBeenCalled();
-    });
+  it('should have ARIA properties for the readonly state', () => {
+    const { container } = render(<NumberInput {...baseProps} isReadOnly />);
+    expect(container.querySelector('input')).toHaveAttribute(
+      'aria-readonly',
+      'true'
+    );
+  });
 
-    it('should keep the same value', () => {
-      expect(input).toHaveProp('value', '1');
-    });
+  it('should call onChange when chaning the value', () => {
+    const onChange = jest.fn(event => {
+      expect(event.target.id).toEqual('some-id');
+      expect(event.target.name).toEqual('some-name');
+      expect(event.target.value).toEqual('47');
+    });
+    const { container } = render(
+      <NumberInput
+        {...baseProps}
+        id="some-id"
+        name="some-name"
+        onChange={onChange}
```

Comment:
The same test in both ways:

```jsx
it('should forward data-attributes', () => {
  const props = createTestProps({ 'data-foo': 'bar' })
  const { container } = render(<NumberInput {...props} />);
  expect(container.querySelector('input')).toHaveAttribute('data-foo', 'bar');
});
```

```js
it('should forward data-attributes', () => {
  const { container } = render(<NumberInput {...baseProps} data-foo="bar" />);
  expect(container.querySelector('input')).toHaveAttribute('data-foo', 'bar');
});
```

I like that no external function is necessary, so the test is easier to comprehend. It is also shorter.

In case one is interested in the full set of props, we could always do

```js
it('should forward data-attributes', () => {
  const props = { ...baseProps, 'data-foo': 'bar' }
  const { container } = render(<NumberInput {...props} />);
  expect(container.querySelector('input')).toHaveAttribute('data-foo', 'bar');
});
```

The only thing to be aware of is that `baseProps` should not define any `jest.fn()` mocks, as they're reused on every test. Maybe the `createTestProps` fn is a better idea for this reason.

Reply:
I guess as createTestProps is all about the defaults (like baseProps) and spreading some more specific things over it doesnt matter much.

Reply:
I didnt mind baseProps over createTestProps. I was „confused“ by the explicit props passing into the component. Instead of just spreading in something.